### ISSUE TITLE
ability to get return type for generic function for a given combination of parameters

### ISSUE THREAD

**Issue by daurnimator** [ziglang/zig#2935]
I think `@typeInfo(func).Fn.return_type` is an optional to represent the `noreturn` case. However, `noreturn` *is* a type: in such a case, the `.return_type` field should just be `noreturn`.

**Comment by andrewrk** [comment]
the problem is generic functions. the return type might not be known if it depends on a parameter for example.

**Comment by daurnimator** [comment]
In that case, could there be a way to get the return type for a given combination of parameters?

**Comment by andrewrk** [comment]
Related: #846

**Comment by yrashk** [comment]
Facing this issue myself, this is a solution I came up with:

```zig
const std = @import("std");

fn gen(t: anytype) @TypeOf(t) {
    return t;
}

pub fn main() void {
    std.debug.print("{}\n", .{@TypeOf(@call(.{}, gen, .{@as(u32, undefined)}))});
}
```

**Comment by InKryption** [comment]
@yrashk Be wary, that doesn't cover all use cases. This, for example, doesn't compile:
```zig
fn gen(t: anytype) @TypeOf(t) {
    return t + 3;
}

fn gen2(v: anytype) @TypeOf(@call(.{}, gen, .{@as(u32, undefined)})) {
    return gen(v);
}

pub fn main() void {
    std.debug.print("{}\n", .{ gen2(@as(u32, undefined)) });
}
```

Because it attempts to use an undefined value at compile-time.

**Comment by yrashk** [comment]
Yeah, I ran into this as well :) I wish it was possible to instantiate these generic functions easily.

**Comment by yrashk** [comment]
If we narrow the scope further, why can't we get a `return_type` of a function whose return type is not deriven comptime? That is, why is this not possible?:

```zig
try expect(@typeInfo(fn (type) bool).Fn.return_type == @TypeOf(bool));
```

Why the return type is an unconditional `anytype`? I am trying to get a bit more familiar with stage1 to (maybe) fix this but I am also wondering if there's something that precludes us from doing so?

cc @andrewrk

**Comment by mlugg** [comment]
The snippet from @InKryption works with modern versions of Zig (with `.{}` changed to `.auto`) because the contents of `@TypeOf` are treated as runtime code. By the way, there's no need to use `@call`: you can just write `gen(@as(u32, undefined))`.

Closing for language simplicity and because `@TypeOf` already presents a reasonable solution to this problem.
